python/taichi/profiler/kernelprofiler.py

```diff
@@ -0,0 +1,207 @@
+
+    def __lt__(self, other):
+        """For sorted()"""
+        return self.total_time < other.total_time
+
+    def insert_record(self, time):
+        if self.counter == 0:
+            self.min_time = time
+            self.max_time = time
+        self.counter += 1
+        self.total_time += time
+        self.min_time = min(self.min_time, time)
+        self.max_time = max(self.max_time, time)
+
+
+class KernelProfiler:
+    """Kernel profiler of Taichi.
+
+    Kernel profiler acquires kernel profiling records from backend, counts records in python scope,
+    and outputs results by printing : :func:`~taichi.profiler.kernelprofiler.KernelProfiler.print_info`.
+    """
+    def __init__(self):
+        """Constructor of class KernelProfiler.
+
+        ``_profiling_mode`` is a boolean value, turn ON/OFF profiler by :func:`~taichi.profiler.kernelprofiler.KernelProfiler.set_kernel_profiler_mode`.
+        ``_total_time_ms`` is a float value, get the value in seconds with :func:`~taichi.profiler.kernelprofiler.KernelProfiler.get_total_time`.
+        ``_traced_records`` is a list of profiling records, acquired from backend by :func:`~taichi.profiler.kernelprofiler.KernelProfiler.update_records`.
+        ``_statistical_results`` is a dict of statistical profiling results, statistics via :func:`~taichi.profiler.kernelprofiler.KernelProfiler.count_results`.
+        """
+        self._profiling_mode = False
+        self._total_time_ms = 0.0
+        self._traced_records = []
+        self._statistical_results = {}
+
+    def set_kernel_profiler_mode(self, mode=False):
+        if type(mode) is bool:
+            self._profiling_mode = mode
+        else:
+            raise TypeError(f'Arg `mode` must be of type boolean. '
+                            f'Type {type(mode)} '
+                            f'is not supported')
+
+    def get_kernel_profiler_mode(self):
+        return self._profiling_mode
+
+    def clear_frontend(self):
+        self._total_time_ms = 0.0
+        self._traced_records.clear()
+        self._statistical_results.clear()
+
+    def get_total_time(self):
+        self.update_records()  # traced records
+        self.count_results()  # _total_time_ms is counted here
+        return self._total_time_ms / 1000  # ms to s
+
+    def query_info(self, name):
+        self.update_records()  # traced records
+        self.count_results()  # statistical results
+        # TODO : query self.StatisticalResult in python scope
+        return impl.get_runtime().prog.query_kernel_profile_info(name)
+
+    def clear_info(self):
+        impl.get_runtime().sync()  #sync first
+        #clear backend & frontend
+        impl.get_runtime().prog.clear_kernel_profile_info()
+        self.clear_frontend()
+
+    def update_records(self):
+        """Acquires profiling records from a backend."""
+        impl.get_runtime().sync()
+        self.clear_frontend()
+        self._traced_records = impl.get_runtime(
```

Comment:
is this formatting intentional..

Reply:
auto formatted by python/taichi/code_format.py